Geometry Spice converted to Goodie: Broken hover and missing calculation feature

bsinger98 opened the pull request:
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.
- [ x] New Instant Answer
  - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
  - [x ] Other: **`New {IA Name} Instant Answer`**
- [ ] Improvement
  - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
  - [ ] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
  - [ ] Other (Role, Template, Test, Documentation, etc.): **`{GoodieRole/Templates/Tests/Docs}: {Short Description}`**
###### Description of changes

This is not ready for production. I need help converting two features.

Feature 1:
When you hover over the diagram the corresponding formula dot should become bold. Currently when you hover over the individual part it becomes bold but the diagram and formulas need to be linked. I wrote a jquery script that would do this in the onShow function but it never seems to get called. I cannot figure out why it is not called.

Feature 2:
I need to convert the calculation feature. If you do calc square a=2, it will display the answer next to all the formulas. Am I required to convert this feature? I was thinking of having an input on the actual page that will have a=[input]. But I need Jquery for this and the same issue popped up with the last feature. 

Original Spice Page: https://duck.co/ia/view/geometry_goodie

Maintainer: @pixunil

daxtheduck commented:
@bsinger98 Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

GuiltyDolphin commented:
Hey @bsinger98 Thanks for this!

Would you mind linking this up to https://github.com/duckduckgo/zeroclickinfo-spice/issues/2463 (instructions are in the issue) so we can effectively track this? It's worth checking the other instructions/guidance if you haven't already; you might find some of it useful.

bsinger98 commented:
Hey @GuiltyDolphin, I read through the link but I wasn't sure how formal I should make the PR because it still has several issues.

moollaza commented on share/goodie/geometry/geometry.json:
```diff
@@ -0,0 +1,14 @@
+{
```
Seems like this file isn't being used anywhere?

moollaza commented on lib/DDG/Goodie/Geometry.pm:
```diff
@@ -0,0 +1,89 @@
+    my $objectSVG = LoadFile(share('svg/' . $remainder . '.svg'));
+    return "plain text response",
+        structured_answer => {
+            id => 'Geometry',
```
You can remove the `id` and `name` here -- we get these from the IA Pages now 👍

bsinger98 replied:
When I remove these lines the build breaks. Should I pull on master and merge the PR with master?

moollaza commented:
@bsinger98 @GuiltyDolphin -- tomorrow I'll be attempting to release the Stopwatch Port. Depending on how that goes we may need to adjust this PR.

But I think what we're going to need to do is make a new IA Page for this, and update the JS/CSS to match the new ID, e.g. `geometry` -> `geometry_goodie`. That will allow me to deploy this to Beta. Then I can deploy the associated Spice removal PR in the Spice repo to get only this Goodie working on beta for testing.

After that, we'll need to change the code here to match the final ID, `geometry` and then I'll be able to merge + release the repos.

GuiltyDolphin commented:
@moollaza Sounds good, I've already updated https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2253 to use the `timer_goodie` IA page - do you know if we have to update the IDs in the source? It seems to have been working fine (same with Stopwatch).

moollaza commented on share/goodie/geometry/subtitle.handlebars:
```diff
@@ -0,0 +1,29 @@
+<div id="zci--geometry">
```
This isn't necessary -- we will wrap the `content` in a div with `class=<id>`, i.e., `<div class=".zci--geometry">`

moollaza commented on share/goodie/geometry/geometry.css:
```diff
@@ -0,0 +1,93 @@
+#zci--geometry-svg {
```
nit: seeing as we wrap the answer in a div, you can use `.zci--geometry` instead

For now though, we need to use the ID `geometry_goodie` because the `geometry` id belongs to the live Spice.

moollaza commented on share/goodie/geometry/geometry.js:
```diff
@@ -0,0 +1,24 @@
+DDH.geometry = DDH.geometry || {};
```
please switch `DDH.geometry` to `DDH.geometry_goodie` -- this will just be temporary so we can test it on beta. Before merging, we'll flip it back so we can seamlessly replace the Spice -- sorry for the extra work!

moollaza commented on t/Geometry.t:
```diff
@@ -0,0 +1,49 @@
+        structured_answer => {
+
+            data => {
+                title    => "My Instant Answer Title",
```
Let's update these test to ensure the hash created and returned in the Perl is correct

moollaza commented:
> do you know if we have to update the IDs in the source?

Yes, the Goodie will need to temporarily update it's CSS/JS to use the ID `geometry_goodie` for now. Once it's tested and working on Beta, we can switch back to the ID `geometry` and merge.

moollaza commented:
@bsinger98 when ready, please also make another PR that removes the Geometry Spice. Once that is ready I can deploy these PRs onto Beta to verify the Goodie implementation works as expected.

bsinger98 commented on share/goodie/geometry/geometry.js:
```diff
@@ -0,0 +1,24 @@
+        return {
+
+            // Function that executes after template content is displayed
+            onShow: function() {
```
@moollaza Do you know why this onShow function does not get called? I can't seem to figure it out.

moollaza replied:
@bsinger98 I suspect it has to do with the Metadata and the Goodie not having its own IA Page. Please create the IA Page for geometry_goodie and I suspect it will start working.

The code looks fine as far as I can tell.

moollaza replied:
Seeing as this isn't a permanent IA Page, I went ahead and created it. I will verify if the IA Page is the problem!

moollaza replied:
@bsinger98 yup, fixed it! 

Your JS should look like this: 

```js
DDH.geometry_goodie = DDH.geometry_goodie || {};

(function(DDH) {
    "use strict";

    DDH.geometry_goodie.build = function(ops) {
        return {
            onShow: function() {
                console.log("look here!");
            }
        };
    };
})(DDH);
```

bsinger98 replied:
Ah cool thanks so much, I was going crazy over this.

moollaza commented:
@bsinger98 Ping! I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

There's absolutely no rush, just wanted to make sure you're not stuck on anything.

bsinger98 commented:
@moollaza Hey sorry I just started a summer programming internship and I have been very busy with that. I am definitely still interested.

edgesince84 commented:
Congrats on the internship @bsinger98!  Glad to hear you're still interested here too :)

bsinger98 commented:
@moollaza Sorry for the long delay. I finished up everything but the test file. I am having an issue testing the formulas part of the response because the IA returns a reference to an Array. Do you know of a way to test this? The test returns got = "ARRAY(0x3006b68)" and expected = 'area'.

moollaza commented on share/goodie/geometry/objectInfo.yml:
```diff
@@ -0,0 +1,39 @@
+---
```
We should make this a hash instead of an array of hashes:

```yml
---
square:
  area: "..."
  perimeter: "..."
  diagonal: "..."
rectangle:
  area:
  perimeter:  
  diagonal:  
...
```

Then you can `LoadFile()`into a hash, which means you don't need to iterate over an array to find a match! (i.e. you can remove `findObject()`) That's the awesome benefit of using YAML with Perl 👍

moollaza commented on share/goodie/geometry/formulas.yml:
```diff
@@ -0,0 +1,18 @@
+volume:
```
we can combine this file into the other YAML file, separated with `---` and slurp them both in together:

```perl
my ($var1, $var2) = LoadFile(share("data.yml"));
```

moollaza commented:
@bsinger98 I took some time to re-write the Perl and Test file. I've simplified the code and opted to use a single YAML file with two sets of info that are loaded into two separate variables. I've updated the test to pass as well.

It looks like we still need to write form the plain text answer, and test it as well.

Here's the code that I ended up with:  https://gist.github.com/moollaza/125876f32b2e7066f4d1a8af1907f51e

I would suggest you update the code accordingly and read it over. Then you'll want to construct the plain text string response, if you're up for it. It's not exactly required, but would be good to have. Note, I switched to use UTF8 chars in the YMAML. You have an `html_enc()` function available that should convert the text to HTML entities for the plain text response.

Thanks for the awesome work so far!

bsinger98 commented:
@moollaza Thanks for all of the help. The way you handled the YAML file was a lot cleaner. I am a little confused about the plain text string response. What do you mean? Are you talking about testing the HTML returned by the Goodie?

bsinger98 commented:
@moollaza I updated the PR to include your code. I changed dataFormula to be a hash instead of an array because the test would sometimes fail due to an order change in the array.

moollaza commented:
> I am a little confused about the plain text string response. What do you mean? Are you talking about testing the HTML returned by the Goodie?

@bsinger98 in the Perl we have `"plain text response"` -- this is actually meant to be a plain-text version of the answer, which is currently used by duckduckgo.com/lite and /html

If possible, it would be good to build a text answer (i.e. provide the formulas: `area = H x W, perimeter = 2(h+w), ...`) and update the tests to check that.

However, this isn't essential and can be done later by yourself or someone else from the community.

moollaza commented on lib/DDG/Goodie/Geometry.pm:
```diff
@@ -18,72 +18,44 @@ zci is_cached => 1;
     
-    # If hash is undefined, means we dont have shape in YML
-    if (!%object) { return; }
+    my $remainder = lc($_);
```
I wasn't thinking earlier, you can `handle remainder_lc` above an then this isn't necessary. End result is the same though 👍

moollaza commented on lib/DDG/Goodie/Geometry.pm:
```diff
@@ -18,72 +18,44 @@ zci is_cached => 1;
-   
-    #print Dumper(${$formulas}{'area'});   
+    my %dataFormula;
+    print Dumper($shape);
```
whoops, let's remove this

bsinger98 commented:
Ok I removed the Dumpers

moollaza commented:
@bsinger98 Deploying to Beta...

bsinger98 commented:
@moollaza What is the status of this PR?

moollaza commented:
@bsinger98 sorry for the long delay here. If you're still interested in finishing this, I'd be happy to keep moving this along.

It seems like the triggering is lacking currently. Can you please port the existing Spice tests to here to ensure there's no regression?

Once we're sure the triggering is good, we'll need to instance of `geometry_goodie` to `geometry`, and then I'll merge this.

I'll then deprecate the Spice when I release this so the `geometry` ID can be associated with the new Goodie files 👍 

Sound good?

moollaza commented:
As well, could you add triggers for `volume of`, `area of`, etc?

We should be able to handle queries like "volume of a cube" or "area of a circle" 👍

bsinger98 commented:
@moollaza I added the triggering and some more test cases, does this look good?

moollaza commented:
@bsinger98 the `calc square` test is failing. That should be `geometry of a square` instead. I'll fix that.

This Thanks! This LGTM :+1:

bsstoner commented:
This was released to production 🚀

bsinger98 commented:
@moollaza Can you add me as a contributor on the page?
